pkg/cmd/boot/boot_upgrade.go

```diff
@@ -0,0 +1,268 @@
+	err := o.Git().FetchBranch(o.Dir, bootConfigURL, "master")
+	if err != nil {
+		return fmt.Errorf("failed to fetch %s", bootConfigURL)
+	}
+
+	currentSha, _, err := o.bootConfigRef(o.Dir, versionStreamURL, versionStreamRef, bootConfigURL)
+	if err != nil {
+		return fmt.Errorf("failed to get boot config ref for version stream: %s", versionStreamRef)
+	}
+	upgradeSha, upgradeVersion, err := o.bootConfigRef(o.Dir, versionStreamURL, "master", bootConfigURL)
+	if err != nil {
+		return fmt.Errorf("failed to get boot config ref for version stream: master")
+	}
+
+	// check if boot config upgrade available
+	if upgradeSha == currentSha {
+		log.Logger().Infof("No boot config upgrade available")
+		return nil
+	}
+
+	log.Logger().Infof("boot config upgrade available!!!!")
+
+	configCloneDir, err := o.cloneBootConfig(bootConfigURL, upgradeVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to clone boot config repo %s", bootConfigURL)
+	}
+
+	err = o.cherryPickCommits(o.Dir, configCloneDir, currentSha, upgradeSha, "upgrade_branch")
+	if err != nil {
+		return errors.Wrap(err, "failed to cherry pick upgrade commits")
+	}
+
+	err = o.excludeFiles(o.Dir, currentSha)
+	if err != nil {
+		return errors.Wrap(err, "failed to exclude files from commit")
+	}
+
+	return nil
+}
+
+func (o *BootUpgradeOptions) bootConfigRef(dir string, versionStreamURL string, versionStreamRef string, configURL string) (string, string, error) {
+	resolver, err := o.CreateVersionResolver(versionStreamURL, versionStreamRef)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to create version resolver %s", configURL)
+	}
+	configVersion, err := resolver.ResolveGitVersion(configURL)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to resolve config url %s", configURL)
+	}
+	cmtSha, err := o.Git().GetCommitPointedToByTag(dir, fmt.Sprintf("v%s", configVersion))
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to get commit pointed to by %s", cmtSha)
+	}
+	return cmtSha, configVersion, nil
+}
+
+func (o *BootUpgradeOptions) cloneBootConfig(configURL string, configRef string) (string, error) {
+	cloneDir, err := ioutil.TempDir("", "")
+	err = os.MkdirAll(cloneDir, util.DefaultWritePermissions)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create directory: %s", cloneDir)
+	}
+
+	err = o.Git().Clone(configURL, cloneDir)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to clone git URL %s to directory %s", configURL, cloneDir)
+	}
+	return cloneDir, nil
+}
+
+func (o *BootUpgradeOptions) cherryPickCommits(dir, cloneDir, fromSha, toSha, branch string) error {
+	cmts := make([]gits.GitCommit, 0)
+	cmts, err := o.Git().GetCommits(cloneDir, fromSha, toSha)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get commits from %s", cloneDir)
+	}
+
+	for i := len(cmts) - 1; i >= 0; i-- {
+		commitSha := cmts[i].SHA
+		log.Logger().Infof("commit sha %s", commitSha)
+
+		err := o.Git().CherryPickTheirs(dir, commitSha)
```

Comment:
What is going to happen if there is a merge conflict after a few cherry-picks but a still few more are left to me picked? For example, 2 cherry-picks are successful, then there is a conflict on the 3rd one but there are still 2 commit left to be cherry-picked. I think we don't have a way to continue in this case after the user will manually solve the conflict.